server/src/diagnostic_ext.rs

```diff
@@ -55,28 +65,38 @@ pub async fn compile_and_publish_diagnostics(
         // Trigger a compilation and get any diagnostics that were reported during it.
         let diagnostics = configuration_set.trigger_compilation();
         // Publish those diagnostics.
-        publish_diagnostics_for_set(client, diagnostics, configuration_set).await;
+        publish_diagnostics_for_set(backend, diagnostics, configuration_set).await;
     }
 }
 
 /// Processes a list of diagnostics and updates the publish map with LSP-compatible diagnostics.
 ///
 /// This function filters out any diagnostics that do not have a span or cannot be converted
 /// to an LSP diagnostic. It then updates the given publish map with the processed diagnostics.
+/// Any diagnostics that do not have a span are returned for further processing.
 pub fn process_diagnostics(
     diagnostics: Vec<slicec::diagnostics::Diagnostic>,
     publish_map: &mut HashMap<Url, Vec<tower_lsp::lsp_types::Diagnostic>>,
-) {
+) -> Vec<slicec::diagnostics::Diagnostic> {
+    let mut spanless_diagnostics = Vec::new();
     diagnostics
         .into_iter()
         .filter_map(|diagnostic| {
+            // The empty span case is handled by the `try_into_lsp_diagnostic` function.
             let span = diagnostic.span()?;
             let uri = convert_slice_url_to_uri(&span.file)?;
-            try_into_lsp_diagnostic(&diagnostic).map(|lsp_diagnostic| (uri, lsp_diagnostic))
+            match try_into_lsp_diagnostic(diagnostic).map(|lsp_diagnostic| (uri, lsp_diagnostic)) {
+                Ok(d) => Some(d),
+                Err(diagnostic) => {
+                    spanless_diagnostics.push(diagnostic);
+                    None
+                }
+            }
```

Comment:
You don't really need the map here because of how simple the `Ok -> Some` case is:
```suggestion
            match try_into_lsp_diagnostic(diagnostic) {
                Ok(lsp_diagnostic) => Some((uri, lsp_diagnostic)),
                Err(diagnostic) => {
                    spanless_diagnostics.push(diagnostic);
                    None
                }
            }
```